Play nice with other middleware for default fields

justanotherdot opened the pull request:
It appears that `absinthe` doesn't like it when you add middleware to a field and will strip the default middleware, which meant for `absinthe_auth` to try to push default middleware back in. The first commit in this PR demonstrates about seven of the fourteen tests failing because we've added a `noop` middleware: this is because `absinthe_auth` is assuming that it should only add default middleware back in if there is no other middleware to be processed. 

This patch should rectify that by checking if the `Absinthe.Resolution` or default middleware are present down the line. If the middleware in the resolution is empty or if either of those aren't present we jam in the middleware. Thankfully, the default middleware is idempotent on resolved resolutions, so this shouldn't cause problems if we add it later to the middleware, although there still may be the case of _others_ doing wonky things to the middleware like how we are doing.

Ideally, we shouldn't even be thinking about this. We should be performing the authorisation and letting the resolution go on it's way if there's nothing for us to resolve. I will try to raise an issue on `absinthe` to see if this can be resolved upstream.

justanotherdot committed:
Check for specific resolvers before defaulting

This checks for the Absinthe.Resolution or default middleware in order
to jam back in the default if it has been stripped. We do not want to
insert the default middleware, however, if we have a function being used
for resolution for a field.

coderdan commented on lib/absinthe_auth/middleware.ex:
```diff
@@ -29,22 +29,23 @@ defmodule AbsintheAuth.Middleware do
-        push_middleware(resolution, {Absinthe.Middleware.MapGet, field})
+  def maybe_append_middleware(%{definition: definition} = resolution) do
+    has_no_resolution_mware? = not Enum.any?(resolution.middleware, fn {mware_mod, _} ->
+      mware_mod == Absinthe.Middleware.MapGet ||
```
Perhaps this could be easier to read with something like:
```elixir
mware_mod in [Absinthe.Middleware.MapGet, Absinthe.Resolution]
```

coderdan commented on lib/absinthe_auth/middleware.ex:
```diff
@@ -29,22 +29,23 @@ defmodule AbsintheAuth.Middleware do
+      mware_mod == Absinthe.Middleware.MapGet ||
+      mware_mod == Absinthe.Resolution
+    end)
+    if Enum.empty?(resolution.middleware) && has_no_resolution_mware? do
```
Do you need to check both of these? The former may be true but have no resolution middleware. So what if there is a non-resolving middleware set but no resolution middleware? The result would be no resolution set.

justanotherdot commented:
I've rebased with some refactoring: 

* Was doing de morgan's law in my head but wasn't considering short-circuit logic so I borked the result
* `has_no_resolution_mware?` is actually wrong because it is two tuples with a two-tuple in the first position, not just a two-tuple.

justanotherdot committed:
Break resolution when other middleware present

This tweaks the movie schema to demonstrate that the `absinthe_auth`
middleware does not play nicely when other middleware is present. This
is because `absinthe` strips the default middleware for fields that have
provided middleware but still need the MapGet default. `absinthe_auth`
was adding this back in, but making the assumption that it should not if
there is other middleware down the line. The following fix is to be more
precise in checking that there are no resolvers in the middleware before
jamming the value back in.

justanotherdot committed:
Set MIX_ENV to test for CI to check PRs
